File: hash_table/0811_subdomain_visit_count/0811_subdomain_visit_count.py
from typing import List
import collections



class Solution:
    def subdomainVisits(self, cpdomains: List[str]) -> List[str]:
        ans = collections.Counter()
        for domain in cpdomains:
            count, domain = domain.split()
            count = int(count)
            frags = domain.split('.')
            for i in range(len(frags)):
                ans[".".join(frags[i:])] += count

        return ["{} {}".format(ct, dom) for dom, ct in ans.items()]


cpdomains = ["9001 discuss.leetcode.com"]
res = Solution().subdomainVisits(cpdomains)
print(res)

File: hash_table/0811_subdomain_visit_count/0811_subdomain_visit_count.go
package main

import (
	"fmt"
	"strconv"
	"strings"
)

func subdomainVisits(cpdomains []string) []string {

	m := make(map[string]int, len(cpdomains)*4)

	for i := range cpdomains {
		cur := strings.Fields(cpdomains[i])
		times, domain := cur[0], cur[1]
		count, _ := strconv.Atoi(times)
		m[domain] += count
		for j := range domain {
			if domain[j] == '.' {
				m[domain[j+1:]] += count
			}
		}
	}

	res := make([]string, 0, len(m))
	for k, v := range m {
		res = append(res, strconv.Itoa(v)+" "+k)
	}

	return res
}

func main() {
	cpdomains := []string{"9001 discuss.leetcode.com"}
	res := subdomainVisits(cpdomains)
	fmt.Println(res)
}
